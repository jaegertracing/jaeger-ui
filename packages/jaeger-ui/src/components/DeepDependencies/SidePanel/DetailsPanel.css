/*
Copyright (c) 2020 The Jaeger Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.Ddg--DetailsPanel {
  background-color: #fafafa;
  border-left: solid 1px #ddd;
  box-shadow: -3px 0 3px rgba(0, 0, 0, 0.1);
  display: flex;
  flex-direction: column;
  height: 100%;
}

.Ddg--DetailsPanel--SvcOpHeader {
  border-bottom: solid 1px #ddd;
  box-shadow: 0px 3px 3px rgba(0, 0, 0, 0.1);
  font-size: 1.5em;
  text-align: right;
  padding: 0.1em 0.3em;
}

.Ddg--DetailsPanel--DecorationHeader {
  font-size: 1.3em;
  text-align: center;
  padding: 0.3em;
}

.Ddg--DetailsPanel--DecorationHeader > span {
  border-bottom: solid 1px #bbb;
  box-shadow: 0px 3px 3px -3px rgba(0, 0, 0, 0.1);
}

.Ddg--DetailsPanel--DetailLink {
  padding-left: 0.2em;
}

.Ddg--DetailsPanel--errorMsg {
  color: #e58c33;
}

.Ddg--DetailsPanel--DetailsCard {
  background-color: #ffffff;
  border: solid 2px rgba(0, 0, 0, 0.3);
  margin: 0.5em;
  overflow: hidden;
}

.Ddg--DetailsPanel--DetailsCard.is-error {
  color: #e58c33;
}

.Ddg--DetailsPanel--LoadingIndicator {
  display: block;
  margin: auto;
  position: absolute;
  left: 0;
  right: 0;
  top: 50%;
  line-height: 0;
}

.Ddg--DetailsPanel--LoadingWrapper {
  flex-grow: 1;
  position: relative;
}

.Ddg--DetailsPanel--PercentCircleWrapper {
  margin: 0 auto;
  max-width: 20vh;
  padding: 0 3%;
}
