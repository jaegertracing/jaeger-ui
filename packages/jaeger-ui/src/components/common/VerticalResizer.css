/*
Copyright (c) 2017 Uber Technologies, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.VerticalResizer {
  left: 0;
  position: absolute;
  right: 0;
  top: 0;
}

.VerticalResizer.is-flipped {
  transform: scaleX(-1);
}

.VerticalResizer--wrapper {
  bottom: 0;
  position: absolute;
  top: 0;
}

.VerticalResizer--dragger {
  border-left: 2px solid transparent;
  cursor: col-resize;
  height: calc(100vh - var(--nav-height));
  margin-left: -1px;
  position: absolute;
  top: 0;
  width: 1px;
}

.VerticalResizer--dragger:hover {
  border-left: 2px solid rgba(0, 0, 0, 0.3);
}

.VerticalResizer.isDraggingLeft > .VerticalResizer--dragger,
.VerticalResizer.isDraggingRight > .VerticalResizer--dragger {
  background: rgba(136, 0, 136, 0.05);
  width: unset;
}

.VerticalResizer.isDraggingLeft > .VerticalResizer--dragger {
  border-left: 2px solid #808;
  border-right: 1px solid #999;
}

.VerticalResizer.isDraggingRight > .VerticalResizer--dragger {
  border-left: 1px solid #999;
  border-right: 2px solid #808;
}

.VerticalResizer--dragger::before {
  position: absolute;
  top: 0;
  bottom: 0;
  left: -8px;
  right: 0;
  content: ' ';
}

.VerticalResizer.isDraggingLeft > .VerticalResizer--dragger::before,
.VerticalResizer.isDraggingRight > .VerticalResizer--dragger::before {
  left: -2000px;
  right: -2000px;
}

.VerticalResizer--gripIcon {
  position: absolute;
  top: 0;
  bottom: 0;
}

.VerticalResizer--gripIcon::before,
.VerticalResizer--gripIcon::after {
  border-right: 1px solid #ccc;
  content: ' ';
  height: 9px;
  position: absolute;
  right: 9px;
  top: 25px;
}

.VerticalResizer--gripIcon::after {
  right: 5px;
}

.VerticalResizer.isDraggingLeft > .VerticalResizer--gripIcon::before,
.VerticalResizer.isDraggingRight > .VerticalResizer--gripIcon::before,
.VerticalResizer.isDraggingLeft > .VerticalResizer--gripIcon::after,
.VerticalResizer.isDraggingRight > .VerticalResizer--gripIcon::after {
  border-right: 1px solid rgba(136, 0, 136, 0.5);
}
