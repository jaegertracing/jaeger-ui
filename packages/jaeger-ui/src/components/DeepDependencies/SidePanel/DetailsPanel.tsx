// Copyright (c) 2020 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import { Tooltip } from 'antd';
import _get from 'lodash/get';
import { connect } from 'react-redux';

import BreakableText from '../../common/BreakableText';
import DetailsCard from '../../common/DetailsCard';
import LoadingIndicator from '../../common/LoadingIndicator';
import NewWindowIcon from '../../common/NewWindowIcon';
import VerticalResizer from '../../common/VerticalResizer';
import JaegerAPI from '../../../api/jaeger';
import extractDecorationFromState, { TDecorationFromState } from '../../../model/path-agnostic-decorations';
import stringSupplant from '../../../utils/stringSupplant';

import { TPathAgnosticDecorationSchema } from '../../../model/path-agnostic-decorations/types';
import { TColumnDefs, TDetails } from '../../common/DetailsCard/types';

import './DetailsPanel.css';

type TProps = TDecorationFromState & {
  decorationSchema: TPathAgnosticDecorationSchema;
  service: string;
  operation?: string | string[] | null;
};

type TState = {
  columnDefs?: TColumnDefs;
  details?: TDetails;
  detailsErred?: boolean;
  detailsLoading?: boolean;
  width?: number;
};

export class UnconnectedDetailsPanel extends React.PureComponent<TProps, TState> {
  state: TState = {};

  componentDidMount() {
    this.fetchDetails();
  }

  componentDidUpdate(prevProps: TProps) {
    if (
      prevProps.operation !== this.props.operation ||
      prevProps.service !== this.props.service ||
      prevProps.decorationSchema !== this.props.decorationSchema
    ) {
      // eslint-disable-next-line react/no-did-update-set-state
      this.setState({
        details: undefined,
        detailsErred: false,
        detailsLoading: false,
      });
      this.fetchDetails();
    }
  }

  fetchDetails() {
    const {
      decorationSchema: {
        detailUrl,
        detailPath,
        detailColumnDefPath,
        opDetailUrl,
        opDetailPath,
        opDetailColumnDefPath,
      },
      operation: _op,
      service,
    } = this.props;

    const operation = _op && !Array.isArray(_op) ? _op : undefined;

    let fetchUrl: string | undefined;
    let getDetailPath: string | undefined;
    let getDefPath: string | undefined;
    if (opDetailUrl && opDetailPath && operation) {
      fetchUrl = stringSupplant(opDetailUrl, { service, operation });
      getDetailPath = stringSupplant(opDetailPath, { service, operation });
      getDefPath = opDetailColumnDefPath && stringSupplant(opDetailColumnDefPath, { service, operation });
    } else if (detailUrl && detailPath) {
      fetchUrl = stringSupplant(detailUrl, { service });
      getDetailPath = stringSupplant(detailPath, { service });
      getDefPath = detailColumnDefPath && stringSupplant(detailColumnDefPath, { service });
    }

    if (!fetchUrl || !getDetailPath) return;

    this.setState({ detailsLoading: true });

    JaegerAPI.fetchDecoration(fetchUrl)
      .then((res: unknown) => {
        let detailsErred = false;
        let details = _get(res, getDetailPath as string);
        if (details === undefined) {
          details = `\`${getDetailPath}\` not found in response`;
          detailsErred = true;
        }
        const columnDefs: TColumnDefs = getDefPath ? _get(res, getDefPath, []) : [];

        this.setState({ columnDefs, details, detailsErred, detailsLoading: false });
      })
      .catch((err: Error) => {
        this.setState({
          details: `Unable to fetch decoration: ${err.message || err}`,
          detailsErred: true,
          detailsLoading: false,
        });
      });
  }

  onResize = (width: number) => {
    this.setState({ width });
  };

  render() {
    const { decorationProgressbar, decorationSchema, decorationValue, operation: _op, service } = this.props;
    const { detailLink } = decorationSchema;
    const { width = 0.3 } = this.state;
    const operation = _op && !Array.isArray(_op) ? _op : undefined;
    return (
      <div className="Ddg--DetailsPanel" style={{ width: `${width * 100}vw` }}>
        <div className="Ddg--DetailsPanel--SvcOpHeader">
          <BreakableText text={service} />
          {operation && <BreakableText text={`::${operation}`} />}
        </div>
        <div className="Ddg--DetailsPanel--DecorationHeader">
          <span>{stringSupplant(decorationSchema.name, { service, operation })}</span>
          {detailLink && (
            <Tooltip arrowPointAtCenter title="More Info">
              <a
                className="Ddg--DetailsPanel--DetailLink"
                href={stringSupplant(detailLink, { service, operation })}
                target="_blank"
                rel="noreferrer noopener"
              >
                <NewWindowIcon />
              </a>
            </Tooltip>
          )}
        </div>
        {decorationProgressbar ? (
          <div className="Ddg--DetailsPanel--PercentCircleWrapper">{decorationProgressbar}</div>
        ) : (
          <span className="Ddg--DetailsPanel--errorMsg">{decorationValue}</span>
        )}
        {this.state.detailsLoading && (
          <div className="Ddg--DetailsPanel--LoadingWrapper">
            <LoadingIndicator className="Ddg--DetailsPanel--LoadingIndicator" />
          </div>
        )}
        {this.state.details && (
          <DetailsCard
            className={`Ddg--DetailsPanel--DetailsCard ${this.state.detailsErred ? 'is-error' : ''}`}
            columnDefs={this.state.columnDefs}
            details={this.state.details}
            header="Details"
          />
        )}
        <VerticalResizer max={0.8} min={0.1} onChange={this.onResize} position={width} rightSide />
      </div>
    );
  }
}

export default connect(extractDecorationFromState)(UnconnectedDetailsPanel);
