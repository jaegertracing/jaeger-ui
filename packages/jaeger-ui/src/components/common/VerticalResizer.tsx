// Copyright (c) 2017 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import cx from 'classnames';

import { TNil } from '../../types';
import DraggableManager, { DraggableBounds, DraggingUpdate } from '../../utils/DraggableManager';

import './VerticalResizer.css';

type VerticalResizerProps = {
  max: number;
  min: number;
  onChange: (newSize: number) => void;
  position: number;
  rightSide?: boolean;
};

type VerticalResizerState = {
  dragPosition: number | TNil;
};

export default class VerticalResizer extends React.PureComponent<VerticalResizerProps, VerticalResizerState> {
  state: VerticalResizerState;

  _dragManager: DraggableManager;
  _rootElm: Element | TNil;

  constructor(props: VerticalResizerProps) {
    super(props);
    this._dragManager = new DraggableManager({
      getBounds: this._getDraggingBounds,
      onDragEnd: this._handleDragEnd,
      onDragMove: this._handleDragUpdate,
      onDragStart: this._handleDragUpdate,
    });
    this._rootElm = undefined;
    this.state = {
      dragPosition: null,
    };
  }

  componentWillUnmount() {
    this._dragManager.dispose();
  }

  _setRootElm = (elm: Element | TNil) => {
    this._rootElm = elm;
  };

  _getDraggingBounds = (): DraggableBounds => {
    if (!this._rootElm) {
      throw new Error('invalid state');
    }
    const { left: clientXLeft, width } = this._rootElm.getBoundingClientRect();
    const { rightSide } = this.props;
    let { min, max } = this.props;
    if (rightSide) [min, max] = [1 - max, 1 - min];
    return {
      clientXLeft,
      width,
      maxValue: max,
      minValue: min,
    };
  };

  _handleDragUpdate = ({ value }: DraggingUpdate) => {
    const dragPosition = this.props.rightSide ? 1 - value : value;
    this.setState({ dragPosition });
  };

  _handleDragEnd = ({ manager, value }: DraggingUpdate) => {
    manager.resetBounds();
    this.setState({ dragPosition: null });
    const dragPosition = this.props.rightSide ? 1 - value : value;
    this.props.onChange(dragPosition);
  };

  render() {
    let left;
    let draggerStyle;
    let isDraggingCls = '';
    const { position, rightSide } = this.props;
    const { dragPosition } = this.state;
    left = `${position * 100}%`;
    const gripStyle = { left };

    if (this._dragManager.isDragging() && this._rootElm && dragPosition != null) {
      isDraggingCls = cx({
        isDraggingLeft: dragPosition < position,
        isDraggingRight: dragPosition > position,
      });
      left = `${dragPosition * 100}%`;
      // Draw a highlight from the current dragged position back to the original
      // position, e.g. highlight the change. Draw the highlight via `left` and
      // `right` css styles (simpler than using `width`).
      const draggerLeft = `${Math.min(position, dragPosition) * 100}%`;
      // subtract 1px for draggerRight to deal with the right border being off
      // by 1px when dragging left
      const draggerRight = `calc(${(1 - Math.max(position, dragPosition)) * 100}% - 1px)`;
      draggerStyle = { left: draggerLeft, right: draggerRight };
    } else {
      draggerStyle = gripStyle;
    }
    return (
      <div
        className={`VerticalResizer ${isDraggingCls} ${rightSide ? 'is-flipped' : ''}`}
        ref={this._setRootElm}
      >
        <div className="VerticalResizer--gripIcon" style={gripStyle} />
        <div
          aria-hidden
          className="VerticalResizer--dragger"
          onMouseDown={this._dragManager.handleMouseDown}
          style={draggerStyle}
        />
      </div>
    );
  }
}
