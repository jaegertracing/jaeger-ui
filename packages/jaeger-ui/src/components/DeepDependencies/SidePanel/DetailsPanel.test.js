// Copyright (c) 2020 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import React from 'react';
import { shallow } from 'enzyme';
import _set from 'lodash/set';

import stringSupplant from '../../../utils/stringSupplant';
import JaegerAPI from '../../../api/jaeger';
import VerticalResizer from '../../common/VerticalResizer';
import { UnconnectedDetailsPanel as DetailsPanel } from './DetailsPanel';

describe('<SidePanel>', () => {
  const service = 'test svc';
  const opString = 'test op';
  const props = {
    decorationSchema: {
      detailUrl: 'http://test.detail.url?someParam=#{service}',
      detailPath: 'test.detail.path',
      opDetailUrl: 'http://test.opDetail.url?someParam=#{service}&otherParam=#{operation}',
      opDetailPath: 'test.opDetail.path',
      name: 'Decorating #{service}',
    },
    decorationValue: 'test decorationValue',
    service,
  };
  const supplantedUrl = stringSupplant(props.decorationSchema.detailUrl, { service });
  const supplantedOpUrl = stringSupplant(props.decorationSchema.opDetailUrl, {
    operation: opString,
    service,
  });
  let fetchDecorationSpy;
  let promise;
  let res;
  let rej;

  beforeAll(() => {
    fetchDecorationSpy = jest.spyOn(JaegerAPI, 'fetchDecoration').mockImplementation(() => {
      promise = new Promise((resolve, reject) => {
        res = resolve;
        rej = reject;
      });
      return promise;
    });
  });

  beforeEach(() => {
    fetchDecorationSpy.mockClear();
  });

  describe('fetchDetails', () => {
    it('fetches correct details url, perferring op-scoped details, or does not fetch at all', async () => {
      const details = 'test details';
      const columnDefs = ['test', 'column', 'defs'];

      const tests = [];

      ['detailUrl#{service}', undefined].forEach(detailUrl => {
        ['detail.path.#{service}', undefined].forEach(detailPath => {
          ['detail.column.def.path.#{service}', undefined].forEach(detailColumnDefPath => {
            ['opDetailUrl#{service}#{operation}', undefined].forEach(opDetailUrl => {
              ['op.detail.path.#{service}.#{operation}', undefined].forEach(opDetailPath => {
                ['op.detail.column.def.path.#{service}', undefined].forEach(opDetailColumnDefPath => {
                  ['op', ['op0', 'op1'], undefined].forEach(operation => {
                    [{ message: 'Err obj with message' }, 'error message', false].forEach(error => {
                      [true, false].forEach(hasDetails => {
                        [true, false].forEach(hasColumnDefPath => {
                          tests.push(async () => {
                            fetchDecorationSpy.mockClear();
                            const detailsPanel = new DetailsPanel({
                              operation,
                              service,
                              decorationSchema: {
                                detailUrl,
                                detailPath,
                                detailColumnDefPath,
                                opDetailUrl,
                                opDetailPath,
                                opDetailColumnDefPath,
                              },
                            });

                            const setStateSpy = jest.spyOn(detailsPanel, 'setState').mockImplementation();
                            detailsPanel.fetchDetails();

                            let supplantedFetchUrl;
                            let supplantedColumnDefPath;
                            let supplantedDetailsPath;

                            if (
                              typeof opDetailUrl === 'string' &&
                              typeof opDetailPath === 'string' &&
                              typeof operation === 'string'
                            ) {
                              supplantedFetchUrl = stringSupplant(opDetailUrl, { service, operation });
                              supplantedDetailsPath = stringSupplant(opDetailPath, { service, operation });
                              if (opDetailColumnDefPath)
                                supplantedColumnDefPath = stringSupplant(opDetailColumnDefPath, {
                                  service,
                                  operation,
                                });
                            } else if (typeof detailUrl === 'string' && typeof detailPath === 'string') {
                              supplantedFetchUrl = stringSupplant(detailUrl, { service });
                              supplantedDetailsPath = stringSupplant(detailPath, { service });
                              if (detailColumnDefPath)
                                supplantedColumnDefPath = stringSupplant(detailColumnDefPath, { service });
                            } else {
                              expect(fetchDecorationSpy).not.toHaveBeenCalled();
                              return;
                            }

                            expect(fetchDecorationSpy).toHaveBeenLastCalledWith(supplantedFetchUrl);
                            expect(setStateSpy).toHaveBeenLastCalledWith({ detailsLoading: true });

                            const expectedSetStateArg = {
                              detailsLoading: false,
                              detailsErred: Boolean(error || !hasDetails),
                            };

                            if (!error) {
                              const result = {};

                              if (hasDetails) {
                                _set(result, supplantedDetailsPath, details);
                                expectedSetStateArg.details = details;
                              } else {
                                expectedSetStateArg.details = `\`${supplantedDetailsPath}\` not found in response`;
                              }

                              if (hasColumnDefPath && supplantedColumnDefPath) {
                                _set(result, supplantedColumnDefPath, columnDefs);
                                expectedSetStateArg.columnDefs = columnDefs;
                              } else {
                                expectedSetStateArg.columnDefs = [];
                              }

                              res(result);
                              await promise;
                              expect(setStateSpy).toHaveBeenLastCalledWith(expectedSetStateArg);
                            } else {
                              const errorMessage = error.message || error;
                              expectedSetStateArg.details = `Unable to fetch decoration: ${errorMessage}`;
                              rej(error);
                              await promise.catch(() => {});
                              expect(setStateSpy).toHaveBeenLastCalledWith(expectedSetStateArg);
                            }
                          });
                        });
                      });
                    });
                  });
                });
              });
            });
          });
        });
      });

      const errors = [];
      await Promise.all(tests.map(test => test().catch(err => errors.push(err))));
      if (errors.length) throw errors;
    });
  });

  describe('render', () => {
    it('renders', () => {
      const wrapper = shallow(<DetailsPanel {...props} />);
      wrapper.setState({ detailsLoading: false });
      expect(wrapper).toMatchSnapshot();
    });

    it('renders with operation', () => {
      const wrapper = shallow(<DetailsPanel {...props} operation={opString} />);
      wrapper.setState({ detailsLoading: false });
      expect(wrapper).toMatchSnapshot();
    });

    it('renders omitted array of operations', () => {
      const wrapper = shallow(<DetailsPanel {...props} operation={['op0', 'op1']} />);
      wrapper.setState({ detailsLoading: false });
      expect(wrapper).toMatchSnapshot();
    });

    it('renders with progressbar', () => {
      const progressbar = <div>stand-in progressbar</div>;
      const wrapper = shallow(<DetailsPanel {...props} decorationProgressbar={progressbar} />);
      wrapper.setState({ detailsLoading: false });
      expect(wrapper).toMatchSnapshot();
    });

    it('renders while loading', () => {
      const wrapper = shallow(<DetailsPanel {...props} />);
      expect(wrapper).toMatchSnapshot();
    });

    it('renders details', () => {
      const wrapper = shallow(<DetailsPanel {...props} />);
      wrapper.setState({ detailsLoading: false, details: 'details string' });
      expect(wrapper).toMatchSnapshot();
    });

    it('renders details error', () => {
      const wrapper = shallow(<DetailsPanel {...props} />);
      wrapper.setState({ detailsLoading: false, details: 'details error', detailsErred: true });
      expect(wrapper).toMatchSnapshot();
    });

    it('renders detailLink', () => {
      const schemaWithLink = {
        ...props.decorationSchema,
        detailLink: 'test details link',
      };
      const wrapper = shallow(<DetailsPanel {...props} decorationSchema={schemaWithLink} />);
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('componentDidMount', () => {
    it('fetches details', () => {
      expect(fetchDecorationSpy).not.toHaveBeenCalled();

      shallow(<DetailsPanel {...props} />);
      expect(fetchDecorationSpy).toHaveBeenCalled();
      expect(fetchDecorationSpy).toHaveBeenLastCalledWith(supplantedUrl);
    });
  });

  describe('componentDidUpdate', () => {
    const expectedState = expect.objectContaining({
      details: undefined,
      detailsErred: false,
      detailsLoading: true,
    });
    let wrapper;

    beforeEach(() => {
      wrapper = shallow(<DetailsPanel {...props} />);
      wrapper.setState({
        details: 'test details',
        detailsErred: false,
        detailsLoading: false,
      });
      expect(fetchDecorationSpy).toHaveBeenCalledTimes(1);
      expect(fetchDecorationSpy).toHaveBeenLastCalledWith(supplantedUrl);
    });

    it('fetches details and clears relevant state if decorationSchema changes', () => {
      const detailUrl = 'http://new.schema.detailsUrl?service=#{service}';
      const newSchema = {
        ...props.decorationSchema,
        detailUrl,
      };
      wrapper.setProps({ decorationSchema: newSchema });
      expect(fetchDecorationSpy).toHaveBeenCalledTimes(2);
      expect(fetchDecorationSpy).toHaveBeenLastCalledWith(stringSupplant(detailUrl, { service }));
      expect(wrapper.state()).toEqual(expectedState);
    });

    it('fetches details and clears relevant state if operation changes', () => {
      wrapper.setProps({ operation: opString });
      expect(fetchDecorationSpy).toHaveBeenCalledTimes(2);
      expect(fetchDecorationSpy).toHaveBeenLastCalledWith(supplantedOpUrl);
      expect(wrapper.state()).toEqual(expectedState);
    });

    it('fetches details and clears relevant state if service changes', () => {
      const newService = 'different test service';
      wrapper.setProps({ service: newService });
      expect(fetchDecorationSpy).toHaveBeenCalledTimes(2);
      expect(fetchDecorationSpy).toHaveBeenLastCalledWith(
        stringSupplant(props.decorationSchema.detailUrl, { service: newService })
      );
      expect(wrapper.state()).toEqual(expectedState);
    });

    it('does nothing if decorationSchema, operation, and service are unchanged', () => {
      wrapper.setProps({ decorationValue: `not_${props.decorationValue}` });
      expect(fetchDecorationSpy).toHaveBeenCalledTimes(1);
    });
  });

  describe('onResize', () => {
    it('updates state', () => {
      const width = 60;
      const wrapper = shallow(<DetailsPanel {...props} />);
      expect(wrapper.state('width')).not.toBe(width);

      wrapper.find(VerticalResizer).prop('onChange')(width);
      expect(wrapper.state('width')).toBe(width);
    });
  });
});
