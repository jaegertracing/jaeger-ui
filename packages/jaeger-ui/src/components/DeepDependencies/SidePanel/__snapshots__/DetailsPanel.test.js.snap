// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SidePanel> render renders 1`] = `
<div
  className="Ddg--DetailsPanel"
  style={
    Object {
      "width": "30vw",
    }
  }
>
  <div
    className="Ddg--DetailsPanel--SvcOpHeader"
  >
    <BreakableText
      className="BreakableText"
      text="test svc"
      wordRegexp={/\\\\W\\*\\\\w\\+\\\\W\\*/g}
    />
  </div>
  <div
    className="Ddg--DetailsPanel--DecorationHeader"
  >
    <span>
      Decorating test svc
    </span>
  </div>
  <span
    className="Ddg--DetailsPanel--errorMsg"
  >
    test decorationValue
  </span>
  <VerticalResizer
    max={0.8}
    min={0.1}
    onChange={[Function]}
    position={0.3}
    rightSide={true}
  />
</div>
`;

exports[`<SidePanel> render renders detailLink 1`] = `
<div
  className="Ddg--DetailsPanel"
  style={
    Object {
      "width": "30vw",
    }
  }
>
  <div
    className="Ddg--DetailsPanel--SvcOpHeader"
  >
    <BreakableText
      className="BreakableText"
      text="test svc"
      wordRegexp={/\\\\W\\*\\\\w\\+\\\\W\\*/g}
    />
  </div>
  <div
    className="Ddg--DetailsPanel--DecorationHeader"
  >
    <span>
      Decorating test svc
    </span>
    <Tooltip
      arrowPointAtCenter={true}
      autoAdjustOverflow={true}
      mouseEnterDelay={0.1}
      mouseLeaveDelay={0.1}
      placement="top"
      title="More Info"
      transitionName="zoom-big-fast"
    >
      <a
        className="Ddg--DetailsPanel--DetailLink"
        href="test details link"
        rel="noreferrer noopener"
        target="_blank"
      >
        <NewWindowIcon
          isLarge={false}
        />
      </a>
    </Tooltip>
  </div>
  <span
    className="Ddg--DetailsPanel--errorMsg"
  >
    test decorationValue
  </span>
  <div
    className="Ddg--DetailsPanel--LoadingWrapper"
  >
    <LoadingIndicator
      centered={false}
      className="Ddg--DetailsPanel--LoadingIndicator"
      small={false}
    />
  </div>
  <VerticalResizer
    max={0.8}
    min={0.1}
    onChange={[Function]}
    position={0.3}
    rightSide={true}
  />
</div>
`;

exports[`<SidePanel> render renders details 1`] = `
<div
  className="Ddg--DetailsPanel"
  style={
    Object {
      "width": "30vw",
    }
  }
>
  <div
    className="Ddg--DetailsPanel--SvcOpHeader"
  >
    <BreakableText
      className="BreakableText"
      text="test svc"
      wordRegexp={/\\\\W\\*\\\\w\\+\\\\W\\*/g}
    />
  </div>
  <div
    className="Ddg--DetailsPanel--DecorationHeader"
  >
    <span>
      Decorating test svc
    </span>
  </div>
  <span
    className="Ddg--DetailsPanel--errorMsg"
  >
    test decorationValue
  </span>
  <DetailsCard
    className="Ddg--DetailsPanel--DetailsCard "
    details="details string"
    header="Details"
  />
  <VerticalResizer
    max={0.8}
    min={0.1}
    onChange={[Function]}
    position={0.3}
    rightSide={true}
  />
</div>
`;

exports[`<SidePanel> render renders details error 1`] = `
<div
  className="Ddg--DetailsPanel"
  style={
    Object {
      "width": "30vw",
    }
  }
>
  <div
    className="Ddg--DetailsPanel--SvcOpHeader"
  >
    <BreakableText
      className="BreakableText"
      text="test svc"
      wordRegexp={/\\\\W\\*\\\\w\\+\\\\W\\*/g}
    />
  </div>
  <div
    className="Ddg--DetailsPanel--DecorationHeader"
  >
    <span>
      Decorating test svc
    </span>
  </div>
  <span
    className="Ddg--DetailsPanel--errorMsg"
  >
    test decorationValue
  </span>
  <DetailsCard
    className="Ddg--DetailsPanel--DetailsCard is-error"
    details="details error"
    header="Details"
  />
  <VerticalResizer
    max={0.8}
    min={0.1}
    onChange={[Function]}
    position={0.3}
    rightSide={true}
  />
</div>
`;

exports[`<SidePanel> render renders omitted array of operations 1`] = `
<div
  className="Ddg--DetailsPanel"
  style={
    Object {
      "width": "30vw",
    }
  }
>
  <div
    className="Ddg--DetailsPanel--SvcOpHeader"
  >
    <BreakableText
      className="BreakableText"
      text="test svc"
      wordRegexp={/\\\\W\\*\\\\w\\+\\\\W\\*/g}
    />
  </div>
  <div
    className="Ddg--DetailsPanel--DecorationHeader"
  >
    <span>
      Decorating test svc
    </span>
  </div>
  <span
    className="Ddg--DetailsPanel--errorMsg"
  >
    test decorationValue
  </span>
  <VerticalResizer
    max={0.8}
    min={0.1}
    onChange={[Function]}
    position={0.3}
    rightSide={true}
  />
</div>
`;

exports[`<SidePanel> render renders while loading 1`] = `
<div
  className="Ddg--DetailsPanel"
  style={
    Object {
      "width": "30vw",
    }
  }
>
  <div
    className="Ddg--DetailsPanel--SvcOpHeader"
  >
    <BreakableText
      className="BreakableText"
      text="test svc"
      wordRegexp={/\\\\W\\*\\\\w\\+\\\\W\\*/g}
    />
  </div>
  <div
    className="Ddg--DetailsPanel--DecorationHeader"
  >
    <span>
      Decorating test svc
    </span>
  </div>
  <span
    className="Ddg--DetailsPanel--errorMsg"
  >
    test decorationValue
  </span>
  <div
    className="Ddg--DetailsPanel--LoadingWrapper"
  >
    <LoadingIndicator
      centered={false}
      className="Ddg--DetailsPanel--LoadingIndicator"
      small={false}
    />
  </div>
  <VerticalResizer
    max={0.8}
    min={0.1}
    onChange={[Function]}
    position={0.3}
    rightSide={true}
  />
</div>
`;

exports[`<SidePanel> render renders with operation 1`] = `
<div
  className="Ddg--DetailsPanel"
  style={
    Object {
      "width": "30vw",
    }
  }
>
  <div
    className="Ddg--DetailsPanel--SvcOpHeader"
  >
    <BreakableText
      className="BreakableText"
      text="test svc"
      wordRegexp={/\\\\W\\*\\\\w\\+\\\\W\\*/g}
    />
    <BreakableText
      className="BreakableText"
      text="::test op"
      wordRegexp={/\\\\W\\*\\\\w\\+\\\\W\\*/g}
    />
  </div>
  <div
    className="Ddg--DetailsPanel--DecorationHeader"
  >
    <span>
      Decorating test svc
    </span>
  </div>
  <span
    className="Ddg--DetailsPanel--errorMsg"
  >
    test decorationValue
  </span>
  <VerticalResizer
    max={0.8}
    min={0.1}
    onChange={[Function]}
    position={0.3}
    rightSide={true}
  />
</div>
`;

exports[`<SidePanel> render renders with progressbar 1`] = `
<div
  className="Ddg--DetailsPanel"
  style={
    Object {
      "width": "30vw",
    }
  }
>
  <div
    className="Ddg--DetailsPanel--SvcOpHeader"
  >
    <BreakableText
      className="BreakableText"
      text="test svc"
      wordRegexp={/\\\\W\\*\\\\w\\+\\\\W\\*/g}
    />
  </div>
  <div
    className="Ddg--DetailsPanel--DecorationHeader"
  >
    <span>
      Decorating test svc
    </span>
  </div>
  <div
    className="Ddg--DetailsPanel--PercentCircleWrapper"
  >
    <div>
      stand-in progressbar
    </div>
  </div>
  <VerticalResizer
    max={0.8}
    min={0.1}
    onChange={[Function]}
    position={0.3}
    rightSide={true}
  />
</div>
`;
